shared/tests/utility/math/TestCoordinates.cpp

```diff
@@ -0,0 +1,1029 @@
+/*
+ * This file is part of the NUbots Codebase.
+ *
+ * The NUbots Codebase is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * The NUbots Codebase is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with the NUbots Codebase.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Copyright 2013 NUbots <nubots@nubots.net>
+ */
+
+#include <Eigen/Core>
+#include <array>
+#include <catch.hpp>
+#include <cfloat>
+#include <cmath>
+#include <fmt/format.h>
+#include <iomanip>
+#include <utility>
+
+#include "utility/math/coordinates.h"
+
+/**
+ * Test data and functions to validate coordinates.cpp functions.
+ *
+ * Test value domain to stay between the values of TEST_DBL_MAX = 5.9667e+153, and TEST_DBL_MIN = 1.4853e-307
+ * This means the usual domain edge cases - DBL_MAX and DBL_MIN - have been left out or replaced. It has been determined
+ * that inputs approaching these edge cases will not be possible in normal operation. See coordinates.h comments for
+ * more details.
+ * @author Liam Craft
+ */
+
+static constexpr double ERROR_THRESHOLD = 1e-6;
+static constexpr double TEST_DBL_MAX    = 5.9667e+153;
+static constexpr double TEST_DBL_MIN    = 1.4853e-307;
```

Comment:
For each of these constants, you should try to get them from limits if you can.
e.g. `std::numeric_limits<double>::max()` and `std::numeric_limits<double>::lowest()` etc.
Also when you're doing tests to make sure floating point error hasn't grown enough instead of using a fixed number like 1e-6 for the error threshold, you should be checking against machine epsilon.

Have a look at https://en.cppreference.com/w/cpp/types/numeric_limits/epsilon
Basically, the error you should expect in a set of floating-point operations is related to the size of the two numbers, as well as how many operations you are doing on them.

Reply:
two questions I have, having read your comment and that page:

1. How would you suggest to take into account how many operations are being done on the floats?
2. In that cppreference example, it has multiplies by a ULP==2. Looking at the [wiki page for ULP](https://en.wikipedia.org/wiki/Unit_in_the_last_place), it has examples which are more like 2^{-51} or 2^{-53}, which makes a lot of sense based on that definition of ULP. Using 2^{-51} where cppreference has used what they're calling `ulp` would certainly cause underflow though, because machine epsilon * 2^{-53} will obviously underflow. Should we disregard that term from the calculation entirely, or simply always use "2" as in their example, or some other approach entirely? It seems like I am probably misunderstanding what cppreference means when it says `ulp`

Reply:
Sometimes it's hard to know exactly how many operations are being done on a float. But what you'll see is that each operation will have a particular error in regards to units in the last place. Most of the time the error will be <= 0.5 ULP. If you look up documentation you can find out what exactly the expected ULP error is for any operation you're doing.

As for what it means in general since you're dealing with floating-point numbers there isn't a single "error" value that you can reliably use. You can think of floating-point numbers being scientific notation with a particular number of decimal places. So every time you do a maths operation the amount of error you gain is related to the relative magnitudes of numbers. So as an example, if you had a "float" that held 5 significant figures you might have 5.0000e10 So checking your error is less than 1e-6 doesn't make sense as the smallest possible number difference would be 5.0001e10. 1e-6 is smaller than the smallest increment you could do to that number. That's what the machine epsilon is designed to help represent, the smallest "step" you can possibly take based on the magnitude of the numbers involved. So when you're checking for error and how accurate your calculations are, you should use epsilon in order to work out how many units in the last place you want to be accurate to. From our example, that's equivalent to how many significant figures you want to be accurate too. Because you're dealing with binary though it's more like how many bits of error you want to be able to have.

This is also why if you're ever doing some maths operations and you have some large numbers and small numbers, sometimes it's beneficial to rearrange your functions so you only ever do operations on numbers of similar magnitude because large and small numbers together can result in numerical instability due to the operation being unable to represent the small changes in a large number.

Reply:
Also as one more example as to where this can get weird. Imagine you were testing a `sin` function and you decided that 1e-6 was a good threshold for checking. Because of how ULP works, if you stuck a number that was smaller than 1e-6 into the sin function (say 1-e10), your error is much much lower than 1e-6 because the relative scales of the two numbers is smaller. Likewise, when you stick a big number into sin your error will be larger

Reply:
Roger that. We'll roll with 2 for now 👍

Reply:
wouldn't that sine example be covered by the `std::fabs(input)` term in their example though?

Reply:
So for a more concrete example if you had a ground truth for v for `sin(1e-10) == v`, your error in those two would be much less than 1e-6. But then if you did `sin(5) == v` you could have a much higher error as the magnitude of the numbers is larger. When you're testing floating point error all you have is ULP because smaller numbers have the same number of ULP error even though they have smaller absolute error.